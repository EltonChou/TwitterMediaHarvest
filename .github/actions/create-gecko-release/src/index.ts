import * as core from '@actions/core'
import * as http from 'http'
import * as https from 'https'
import semver from 'semver'

type Inputs = {
  endpoint: string
  id: string
  version: string
  url: string
  minBrowserVersion?: string
  apiKey: string
}

type ReleaseBrowserSpec = { min_version?: string }
type CreateReleaseBody = {
  id: string
  version: string
  url: string
  browser?: ReleaseBrowserSpec
}

function getInputs(): Inputs {
  return {
    endpoint: core.getInput('endpoint', {
      required: true,
    }),
    id: core.getInput('id', { required: true }),
    version: core.getInput('version', { required: true }),
    url: core.getInput('url', { required: true }),
    minBrowserVersion:
      core.getInput('min-browser-version', {
        required: false,
      }) || undefined,
    apiKey: core.getInput('api-key', { required: true }),
  }
}

function validateInputs(inputs: Inputs) {
  if (!inputs.endpoint) throw new Error('endpoint is required')

  if (!URL.canParse(inputs.endpoint))
    throw new Error('endpoint is not a valid URL')

  if (!inputs.id) throw new Error('id is required')

  if (!inputs.version || !semver.valid(inputs.version))
    throw new Error('version is not a valid semver')

  if (!inputs.url) throw new Error('url is required')

  if (inputs.minBrowserVersion && !semver.coerce(inputs.minBrowserVersion))
    throw new Error('min-browser-version is not a valid semver')
}

function postJson(
  endpoint: string,
  payload: Record<string, unknown>,
  apiKey: string
): Promise<{ statusCode?: number; body: string }> {
  return new Promise((resolve, reject) => {
    const url = new URL(endpoint)
    const data = JSON.stringify(payload)
    const options: http.RequestOptions = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(data),
        'X-Api-Key': apiKey,
      },
    }
    const lib = url.protocol === 'https:' ? https : http
    const req = lib.request(url, options, (res: http.IncomingMessage) => {
      let body = ''
      res.on('data', (chunk: Buffer) => (body += chunk.toString()))
      res.on('end', () => resolve({ statusCode: res.statusCode, body }))
    })
    req.on('error', reject)
    req.write(data)
    req.end()
  })
}

async function run() {
  try {
    const inputs = getInputs()
    validateInputs(inputs)

    const payload: CreateReleaseBody = {
      id: inputs.id,
      version: inputs.version,
      url: inputs.url,
    }
    if (inputs.minBrowserVersion) {
      payload.browser = { min_version: inputs.minBrowserVersion }
    }

    const res = await postJson(inputs.endpoint, payload, inputs.apiKey)
    if (!res.statusCode || Math.floor(res.statusCode / 100) !== 2) {
      core.setFailed(`Failed to create release: ${res.statusCode}`)
      core.error(res.body)
      return
    }

    core.info('Release created successfully')
    core.setOutput('response', res.body)
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err)
    core.setFailed(message)
  }
}

run()
