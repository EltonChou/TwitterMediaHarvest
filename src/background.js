/**
 * listen to message from content script.
 */
// eslint-disable-next-line no-undef
chrome.runtime.onMessage.addListener(async request => {
  !request.medias
    ? await downloadVideo(JSON.parse(request.info))
    : downloadImage(JSON.parse(request.info), JSON.parse(request.medias))
})

/**
 * trigger browser-download
 * @param {JSON} images images-data
 */
function downloadImage(info, medias) {
  for (const media of medias) {
    // eslint-disable-next-line no-undef
    chrome.downloads.download({
      url: media.url,
      filename: `${info.screenName}-${info.userId}-${info.tweetId}-${
        media.name
      }`,
      conflictAction: 'overwrite',
    })
  }
}

function downloadVideo(info) {
  let regex = new RegExp('\\w+\\.mp4')
  fetchMedia(info, url => {
    const name = url.match(regex)
    // eslint-disable-next-line no-undef
    chrome.downloads.download({
      url: url,
      filename: `${info.screenName}-${info.userId}-${info.tweetId}-${name}`,
      conflictAction: 'overwrite',
    })
  })
}

function fetchMedia(info, fn) {
  // eslint-disable-next-line no-undef
  const header = new Headers()
  const url = `https://api.twitter.com/2/timeline/conversation/${
    info.tweetId
  }.json?tweet_mode=extended`
  const activateUrl = 'https://api.twitter.com/1.1/guest/activate.json'

  header.append(
    'Authorization',
    'Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA'
  )
  header.append(
    'User-Agent',
    'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1'
  )
  header.append('cache-control', 'no-cache')

  // eslint-disable-next-line no-undef
  chrome.cookies.get(
    { url: 'https://twitter.com', name: 'ct0' },
    async cookie => {
      header.append('x-csrf-token', cookie.value)

      // eslint-disable-next-line no-undef
      let res = await fetch(activateUrl, {
        method: 'POST',
        mode: 'cors',
        credentials: 'same-origin',
        headers: header,
      })

      let data = await res.json()

      header.append('x-guest-token', data['guest_token'])

      // eslint-disable-next-line no-undef
      let mediaRes = await fetch(url, {
        method: 'GET',
        credentials: 'same-origin',
        headers: header,
      })

      let conversation = await mediaRes.json()
      // console.log(mediaUrl)
      let variants =
        conversation.globalObjects.tweets[info.tweetId].extended_entities.media[
          '0'
        ].video_info.variants

      let hd = 1
      let targetId = 0
      for (let i in variants) {
        if (variants[i].bitrate) {
          if (variants[i].bitrate > hd) {
            hd = variants[i].bitrate
            targetId = i
          }
        }
      }
      fn(variants[targetId].url)
    }
  )
}
