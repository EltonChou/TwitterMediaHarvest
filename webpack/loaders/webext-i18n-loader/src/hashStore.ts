import crypto from 'crypto'
import { validate } from 'schema-utils'

/**
 * @typedef {Object} HashOptions
 * @property {'sha256' | string} alg
 */
export type HashOptions = {
  alg: 'sha256' | string
}

export class HashStore {
  protected options: HashOptions
  readonly contextHash: Map<string, string>
  readonly msgIdHash: Map<string, string>

  constructor(options: HashOptions) {
    validate(
      {
        type: 'object',
        properties: {
          alg: {
            type: 'string',
          },
        },
        additionalProperties: true,
      },
      options
    )
    this.options = options
    this.contextHash = new Map()
    this.msgIdHash = new Map()
  }

  digsetContext(context: string) {
    return crypto.hash(this.options.alg, context, 'hex').slice(0, 7)
  }

  digsetMsgId(msgId: string) {
    return crypto.hash(this.options.alg, msgId, 'hex').slice(0, 10)
  }

  /**
   * @returns digested hash
   */
  saveContext(context: string) {
    const digestedHash = this.contextHash.get(context)
    if (digestedHash) return digestedHash

    const hash = this.digsetContext(context)
    this.contextHash.set(context, hash)
    return hash
  }

  getContextHash(context: string) {
    return this.contextHash.get(context)
  }

  /**
   * @returns digested hash
   */
  saveMsgId(msgId: string) {
    const digestedHash = this.msgIdHash.get(msgId)
    if (digestedHash) return digestedHash

    const hash = this.digsetMsgId(msgId)
    this.msgIdHash.set(msgId, hash)
    return hash
  }

  getMsgIdHash(msgId: string) {
    return this.msgIdHash.get(msgId)
  }
}

export const getHashStore = (() => {
  let store: HashStore
  return () => (store ??= new HashStore({ alg: 'sha256' }))
})()
