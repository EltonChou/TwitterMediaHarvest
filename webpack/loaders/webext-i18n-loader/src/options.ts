import type { LoaderContext } from 'webpack'

export type LoaderOptions = {
  expressions: string | RegExp | (RegExp | string)[]
}

export const getOptions = (
  loaderContext: LoaderContext<LoaderOptions>
): LoaderOptions =>
  loaderContext.getOptions({
    type: 'object',
    properties: {
      expressions: {
        anyOf: [
          { type: 'array' },
          { type: 'string' },
          { instanceof: 'RegExp' },
        ],
      },
    },
    additionalProperties: false,
  })
