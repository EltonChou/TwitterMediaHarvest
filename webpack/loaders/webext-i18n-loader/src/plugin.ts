import { type HashStore, getHashStore } from './hashStore'
import fs from 'fs/promises'
import { glob } from 'glob'
import path from 'path'
import { validate } from 'schema-utils'
import { Compilation, sources } from 'webpack'
import type { Compiler, WebpackPluginInstance } from 'webpack'

const { RawSource } = sources

type Translation = {
  context?: string
  msgId: string
  content: string
}

type WebextTranslations = {
  [msgId: string]: WebextLocalMessage
}

type WebextLocalMessage = {
  message: string
}

export interface WebextI18nPluginOptions {
  poDir: string
  rawContexts?: string | RegExp | (string | RegExp)[]
  /**
   * If context is matched in {@link rawContexts},
   * `msg` and `context` will be passed as raw values instead of hash values.
   *
   * default maker:
   * @example
   * const defaultMaker = (msg: string, context?: string) => context + '_' + msg
   *
   * @param {string} msgId
   * @param {string} context
   * @returns {string} Output should only contains `[a-z][A-Z][0-9]` and `_`
   */
  messageIdMaker?: (msgId: string, context?: string) => string
}

export class WebextI18nPlugin implements WebpackPluginInstance {
  readonly hashStore: HashStore

  constructor(readonly options: WebextI18nPluginOptions) {
    validate(
      {
        type: 'object',
        required: ['poDir'],
        additionalProperties: false,
        properties: {
          poDir: { type: 'string' },
          rawContexts: {
            oneOf: [
              {
                type: 'array',
                items: {
                  anyof: [{ type: 'string' }, { instanceof: 'RegExp' }],
                },
              },
              { type: 'string' },
              { instanceof: 'RegExp' },
            ],
          },
          messageIdMaker: {
            instanceof: 'Function',
          },
        },
      },
      options
    )
    this.options = options
    this.hashStore = getHashStore()
  }

  isRawContext(context?: string): boolean {
    if (this.options.rawContexts === undefined || context === undefined)
      return false
    return Array.isArray(this.options.rawContexts)
      ? this.options.rawContexts.some(
          rawContext => context.match(rawContext) !== null
        )
      : context.match(this.options.rawContexts) !== null
  }

  makeMsgId(msgId: string, context?: string): string {
    if (this.options.messageIdMaker)
      return this.options.messageIdMaker(msgId, context)

    return context ? context + '_' + msgId : msgId
  }

  transformToLocaleMsgId(hashStore: HashStore) {
    return (msgId: string, context?: string): string => {
      if (context === undefined || context === null)
        return this.makeMsgId(hashStore.getMsgIdHash(msgId) ?? msgId)

      if (this.isRawContext(context)) {
        return this.makeMsgId(msgId, context)
      } else {
        const ctxHash = hashStore.getContextHash(context)
        const msgIdHash = hashStore.getMsgIdHash(msgId)
        return this.makeMsgId(msgIdHash ?? msgId, ctxHash)
      }
    }
  }

  apply(compiler: Compiler) {
    const pluginName = this.constructor.name
    const hashStore = this.hashStore

    compiler.hooks.thisCompilation.tap({ name: pluginName }, compilation => {
      const logger = compilation.getLogger('webext-i18n-plugin')
      // const cache = compilation.getCache(pluginName)

      compilation.hooks.processAssets.tapPromise(
        {
          name: pluginName,
          stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
          additionalAssets: true,
        },
        async _unusedAssets => {
          logger.log(`Finding po files from ${this.options.poDir}`)
          const localePoMap = new Map<string, string>()
          const poFiles = await glob(path.resolve(this.options.poDir, '*.po'))

          compilation.fileDependencies.addAll(poFiles)

          for (const file of poFiles) {
            const locale = parseLocaleName(file)
            localePoMap.set(locale, file)
          }

          for (const [locale, poFile] of localePoMap.entries()) {
            logger.log(`Parsing locale: ${locale} from ${poFile}`)

            const translations = await readTranslationsFromPo(poFile)
            const webextTranslationEntries = translations.map<
              [string, WebextLocalMessage]
            >(translation => [
              this.transformToLocaleMsgId(hashStore)(
                translation.msgId,
                translation.context
              ),
              { message: translation.content },
            ])
            const webextTranslations: WebextTranslations = Object.fromEntries(
              webextTranslationEntries
            )

            const outputFile = path.join('_locales', locale, 'messages.json')

            logger.log(`locale ${locale}: ${poFile} => ${outputFile}`)
            compilation.emitAsset(
              outputFile,
              new RawSource(JSON.stringify(webextTranslations))
            )
          }
        }
      )
    })
  }
}

function parseLocaleName(poFilename: string) {
  return path.basename(poFilename, '.po')
}

async function readTranslationsFromPo(poFile: string): Promise<Translation[]> {
  const poContent = await fs.readFile(path.resolve(poFile))
  const { po } = await import('gettext-parser')
  const content = po.parse(poContent)

  const translations: Translation[] = []
  for (const contextObject of Object.values(content.translations)) {
    for (const [msgId, translation] of Object.entries(contextObject)) {
      if (msgId === '') break
      translations.push({
        content: translation.msgstr.at(0) ?? '',
        msgId: translation.msgid,
        context: translation.msgctxt,
      })
    }
  }

  return translations
}
