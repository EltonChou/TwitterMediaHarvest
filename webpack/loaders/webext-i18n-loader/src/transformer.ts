import { getHashStore } from './hashStore'
import { LoaderOptions } from './options'
import ts from 'typescript'

function replaceContextId<T extends ts.Expression>(
  expression: T,
  transformerCtx: ts.TransformationContext
) {
  if (ts.isStringLiteral(expression)) {
    const digestedHash = getHashStore().saveContext(expression.text)

    return transformerCtx.factory.createStringLiteral(digestedHash)
  }

  return expression
}

function replaceMsgId<T extends ts.Expression>(
  expression: T,
  transformerCtx: ts.TransformationContext
) {
  if (ts.isStringLiteral(expression)) {
    const digestedHash = getHashStore().saveMsgId(expression.text)

    return transformerCtx.factory.createStringLiteral(digestedHash)
  }

  return expression
}

function replaceI18nArguments(
  expr: ts.CallExpression,
  ctx: ts.TransformationContext
) {
  const [msgIdArg, contextArg, placeholders] = expr.arguments

  if (
    msgIdArg !== undefined &&
    contextArg !== undefined &&
    placeholders !== undefined
  ) {
    return ts.factory.createCallExpression(
      ts.factory.createIdentifier(expr.expression.getText()),
      undefined,
      [
        replaceMsgId(msgIdArg, ctx),
        replaceContextId(contextArg, ctx),
        placeholders,
      ]
    )
  }

  if (msgIdArg !== undefined && contextArg !== undefined) {
    return ts.factory.createCallExpression(
      ts.factory.createIdentifier(expr.expression.getText()),
      undefined,
      [replaceMsgId(msgIdArg, ctx), replaceContextId(contextArg, ctx)]
    )
  }

  if (msgIdArg !== undefined) {
    return ts.factory.createCallExpression(
      ts.factory.createIdentifier(expr.expression.getText()),
      undefined,
      [replaceMsgId(msgIdArg, ctx)]
    )
  }

  return expr
}

function isI18nCallExpression(
  expression: LoaderOptions['expressions']
): (node: ts.CallExpression) => boolean {
  return Array.isArray(expression)
    ? node =>
        expression.some(expr => node.expression.getText().match(expr) !== null)
    : node => node.expression.getText().match(expression) !== null
}

type TransformerOptions = {
  config: LoaderOptions
}

export const transformer =
  (options: TransformerOptions): ts.TransformerFactory<ts.SourceFile> =>
  ctx =>
  node => {
    const visitChild: ts.Visitor<ts.Node, ts.Node> = node => {
      if (
        ts.isCallExpression(node) &&
        isI18nCallExpression(options.config.expressions)(node)
      ) {
        const newExpr = replaceI18nArguments(node, ctx)
        if (newExpr) return newExpr
      }
      return ts.visitEachChild(node, visitChild, ctx)
    }

    const visitNode: ts.Visitor<ts.Node, ts.Node> = node =>
      ts.visitEachChild(node, visitChild, ctx)

    return ts.visitNode(node, visitNode, ts.isSourceFile)
  }
