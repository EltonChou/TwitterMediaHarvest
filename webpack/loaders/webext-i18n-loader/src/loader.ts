import type { LoaderOptions } from './options.js'
import { getOptions } from './options.js'
import { transformer } from './transformer.js'
import path from 'path'
import ts from 'typescript'
import type { LoaderDefinitionFunction } from 'webpack'

/**
 * TODO: Try to make this loader can be placed in front or behind `ts-loader`.
 * If the loader is placed in front, it should transforme code only.
 * If the loader is placed behind, it should transform code and update sourcemap.
 **/
const loader: LoaderDefinitionFunction<LoaderOptions> = function loader(
  content
) {
  const callback = this.async()
  const logger = this.getLogger('i18n-loader')
  const options = getOptions(this)

  const tsConfig = ts.readConfigFile(
    path.resolve(process.cwd(), 'tsconfig.json'),
    ts.sys.readFile
  )

  if (tsConfig.error) {
    logger.error(tsConfig.error)
    return callback(new Error('Failed to load typescript configuration file.'))
  }

  const srcFile = ts.createSourceFile(
    'temp',
    content,
    tsConfig.config.compilerOptions.target,
    true
  )
  const result = ts.transform([srcFile], [transformer({ config: options })])
  const transformedContent = result.transformed.at(0)
  const outputContent = transformedContent
    ? ts.createPrinter().printFile(transformedContent)
    : content

  return callback(null, outputContent)
}

export default loader
