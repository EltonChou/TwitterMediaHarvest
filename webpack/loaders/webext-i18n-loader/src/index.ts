import loader from './loader'

export { WebextI18nPlugin } from './plugin'
export { transformer } from './transformer'

export default loader
