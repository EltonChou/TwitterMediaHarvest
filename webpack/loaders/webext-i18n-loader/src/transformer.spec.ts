import { transformer } from '../src/transformer'
import fs from 'node:fs/promises'
import path from 'node:path'
import ts from 'typescript'

test('transformer', async () => {
  const testFile = path.normalize(
    path.resolve(path.dirname(__filename), '..', 'fixtures', 'i18n.ts')
  )

  const content = await fs.readFile(testFile, { encoding: 'utf-8', flag: 'r' })

  const srcFile = ts.createSourceFile(
    'temp',
    content,
    ts.ScriptTarget.ESNext,
    true
  )
  const result = ts.transform(
    [srcFile],
    [transformer({ config: { expressions: ['i18n', '_', 'gettext'] } })]
  )
  const [transformedContent] = result.transformed

  expect(transformedContent).toBeDefined()

  const outputContent = transformedContent
    ? ts.createPrinter().printFile(transformedContent)
    : content

  expect(outputContent).toMatchSnapshot('transformed')
})
