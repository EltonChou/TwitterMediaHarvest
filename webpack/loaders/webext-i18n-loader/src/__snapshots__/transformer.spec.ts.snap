// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer: transformed 1`] = `
"function i18n(msg: string, context?: string, placeholders?: Record<string, string>) {
    return msg + context || '' + JSON.stringify(placeholders);
}
function nope(msg: string) {
    return msg;
}
function getText(msg: string) {
    return msg;
}
const enum Context {
    TEST = 'test'
}
i18n("6a30f630de");
i18n("6a30f630de", Context.TEST);
i18n("6a30f630de", "9f86d08", { foo: 'bar' });
nope('kappa');
getText('keepo');
"
`;
